torch/lib/c10d/ProcessGroupNCCL.hpp

```diff
@@ -125,9 +132,62 @@ class ProcessGroupNCCL : public ProcessGroup {
     // to the store.
     std::shared_ptr<Store> store_;
 
+    // Store a Future work associated with WorkNCCL.
+    c10::intrusive_ptr<c10::ivalue::Future> futureWork_;
+
     friend class ProcessGroupNCCL;
   };
 
+  // FutureNCCL is a subclass of c10d ivalue's Future. FutureNCCL has a
+  // reference to WorkNCCL and NCCL collective's outputs. The goal is to
+  // use this class in getFuture API of WorkNCCL. Once NCCL kernel is done
```

Comment:
> Once NCCL kernel is done executing, this future will have NCCL collective's outputs as its value

This isn't true? The future has a value even before the kernel is done executing?

Add a comment here mentioning this Future is mostly a wrapper to synchronize streams appropriately and it mostly enabled the async programming model of CUDA while trying to adhere to the Future interface.